src/app/core/services/milestone.service.ts

```diff
@@ -47,4 +47,42 @@ export class MilestoneService {
     milestoneData.push(Milestone.DefaultMilestone);
     return milestoneData;
   }
+
+  /**
+   * Gets the open milestone with the earliest deadline.
+   * Returns null if there is no open milestone with deadline.
+   */
+  getEarliestOpenMilestone(): Milestone {
+    let earliestOpenMilestone: Milestone = null;
+    for (const milestone of this.milestones) {
+      if (!milestone.deadline || milestone.state !== 'open') {
+        continue;
+      }
+      if (earliestOpenMilestone === null) {
+        earliestOpenMilestone = milestone;
+      } else if (milestone.deadline < earliestOpenMilestone.deadline) {
+        earliestOpenMilestone = milestone;
+      }
+    }
+    return earliestOpenMilestone;
+  }
+
+  /**
+   * Gets the closed milestone with the latest deadline.
+   * Returns null if there is no closed milestone with deadline.
+   */
+  getLatestClosedMilestone(): Milestone {
+    let latestClosedMilestone: Milestone = null;
+    for (const milestone of this.milestones) {
+      if (!milestone.deadline || milestone.state !== 'closed') {
+        continue;
+      }
+      if (latestClosedMilestone === null) {
+        latestClosedMilestone = milestone;
+      } else if (milestone.deadline > latestClosedMilestone.deadline) {
+        latestClosedMilestone = milestone;
+      }
+    }
+    return latestClosedMilestone;
+  }
```

Comment:
Functions are well written, documented and accomplishes its task in a simple way. 
Perfect representation of KISS.